File: paleo_workbench/ui/workstation/keybinding_manager.py
"""全键盘编图流管理器（M5）+ 状态提示条。

键处理走 **事件过滤器**（composite 与 canvas 双挂）：offscreen 下 QShortcut
模拟键不可靠（仓库既有经验），且 QAbstractItemView/画布会吞未知按键不透传
父级；过滤器对两类接收者都能先见先得。文本输入聚焦时全部让路（复用
:func:`paleo_workbench.ui.shortcuts.focus_in_text_input`，00-decisions D5）。

快捷键面（D5 冲突表已核对全仓占用）：
- ``Space`` 按住临时平移（press 进 pan / release 还原原工具，FSM 瞬态）；
- ``Tab`` 循环选中当前层要素（仅 DIGITIZING/ADJUSTING，不抢焦点链语义）；
- ``Z`` / ``X`` 画布中心放大/缩小（×1.5，无修饰键时才生效，避免吞 Ctrl+Z）；
- ``Ctrl+D`` 吸取选中要素相带属性装备画刷（吸色管的键盘路径，D10）；
- ``Esc`` 安全退出链：退出当前工具 → 关闭浮面向导 → IDLE（无手势期接管；
  手势期的取消仍由既有 map cancel 动作先行，见 04 #23）。
"""
from __future__ import annotations

from typing import Any

from PySide6.QtCore import QEvent, QObject, Qt
from PySide6.QtWidgets import QLabel, QWidget

from paleo_workbench.ui.workstation.mode_state import (
    ModeEvent,
    WorkstationMode,
)

#: 中心缩放倍率。
ZOOM_STEP = 1.5


class KeybindingHintBar(QLabel):
    """画布底部常驻提示条：文本随 FSM 模式切换（信号驱动，S5-2）。"""

    def __init__(self, parent: QWidget | None = None) -> None:
        super().__init__(parent)
        self.setObjectName("KeybindingHintBar")
        self.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self.setToolTip("编图快捷键随当前模式变化")

    def apply_mode(self, mode: WorkstationMode) -> None:
        from paleo_workbench.ui.workstation.mode_state import MODE_HINTS

        self.setText(MODE_HINTS.get(mode, ""))


class WorkstationKeyBindingManager(QObject):
    """composite/canvas 双挂键过滤器 + Esc 退出链 + FSM 事件桥。"""

    def __init__(self, composite: Any, parent: QObject | None = None) -> None:
        super().__init__(parent)
        self._composite = composite
        self._pan_restore_tool: str | None = None
        self._last_tool_id: str | None = None

    # -- 安装 -----------------------------------------------------------------

    def install(self) -> None:
        composite = self._composite
        composite.installEventFilter(self)
        canvas = getattr(composite, "canvas", None)
        if canvas is not None:
            canvas.installEventFilter(self)
        qc_hub = getattr(composite, "qc_hub", None)
        if qc_hub is not None:
            qc_hub.installEventFilter(self)
        # 工具变化 → FSM（临时平移期抑制，防止模式闪烁）。
        controller = getattr(composite, "edit_controller", None)
        if controller is not None:
            controller.state_changed.connect(self._on_tool_state_changed)

    # -- 工具态 → FSM -----------------------------------------------------------

    def _on_tool_state_changed(self, *_args) -> None:
        if self._pan_restore_tool is not None:
            return  # 临时平移中的工具切换不是用户模式事件
        controller = self._composite.edit_controller
        tool = getattr(controller.tools, "active_tool", None)
        tool_id = str(getattr(tool, "tool_id", "") or "")
        machine = getattr(self._composite, "mode_state", None)
        if machine is None:
            return
        if tool_id and tool_id != "pan":
            if tool_id != self._last_tool_id:
                self._last_tool_id = tool_id
                machine.dispatch_tool_activated(tool_id)
        else:
            if self._last_tool_id:
                self._last_tool_id = None
                machine.dispatch(ModeEvent.TOOL_DEACTIVATED)

    # -- 事件过滤 ---------------------------------------------------------------

    def eventFilter(self, obj, event) -> bool:  # noqa: N802
        from paleo_workbench.ui.shortcuts import focus_in_text_input

        composite = self._composite
        if obj is getattr(composite, "qc_hub", None):
            if event.type() == QEvent.Type.Show:
                composite.mode_state.dispatch(ModeEvent.QC_HUB_ACTIVATED)
            elif event.type() == QEvent.Type.Hide:
                composite.mode_state.dispatch(ModeEvent.QC_HUB_CLOSED)
            return False

        if event.type() == QEvent.Type.KeyPress:
            if focus_in_text_input():
                return False
            key = event.key()
            modifiers = event.modifiers()
            if key == Qt.Key.Key_Space and modifiers == Qt.KeyboardModifier.NoModifier:
                self.begin_temporary_pan()
                return True
            if key == Qt.Key.Key_Tab:
                if self._cycle_scope_active():
                    self.cycle_selection()
                    return True
                return False
            if key == Qt.Key.Key_Z and modifiers == Qt.KeyboardModifier.NoModifier:
                self.zoom_center(ZOOM_STEP)
                return True
            if key == Qt.Key.Key_X and modifiers == Qt.KeyboardModifier.NoModifier:
                self.zoom_center(1.0 / ZOOM_STEP)
                return True
            if key == Qt.Key.Key_D and (
                modifiers & Qt.KeyboardModifier.ControlModifier
            ):
                self.pick_facies_from_selection()
                return True
            if key == Qt.Key.Key_Escape:
                self.handle_escape()
                return True
        elif event.type() == QEvent.Type.KeyRelease:
            if event.key() == Qt.Key.Key_Space and self._pan_restore_tool is not None:
                self.release_temporary_pan()
                return True
        return False

    def _cycle_scope_active(self) -> bool:
        machine = getattr(self._composite, "mode_state", None)
        if machine is None:
            return False
        return machine.mode in (WorkstationMode.DIGITIZING,
                                WorkstationMode.ADJUSTING_BOUNDARY)

    # -- 动作 -------------------------------------------------------------------

    def begin_temporary_pan(self) -> None:
        controller = self._composite.edit_controller
        tool = getattr(controller.tools, "active_tool", None)
        tool_id = str(getattr(tool, "tool_id", "") or "")
        if tool_id in ("", "pan"):
            return
        self._pan_restore_tool = tool_id
        machine = getattr(self._composite, "mode_state", None)
        if machine is not None:
            machine.dispatch(ModeEvent.PAN_HELD)
        controller.activate_tool("pan")

    def release_temporary_pan(self) -> None:
        restore = self._pan_restore_tool
        self._pan_restore_tool = None
        machine = getattr(self._composite, "mode_state", None)
        if machine is not None:
            machine.dispatch(ModeEvent.PAN_RELEASED)
        if restore:
            self._composite.edit_controller.activate_tool(restore)

    def cycle_selection(self) -> None:
        handler = getattr(self._composite, "tab_cycle_selection", None)
        if callable(handler):
            handler()

    def zoom_center(self, factor: float) -> None:
        canvas = getattr(self._composite, "canvas", None)
        if canvas is not None:
            canvas.zoom_by(float(factor))

    def pick_facies_from_selection(self) -> None:
        handler = getattr(self._composite, "ctrl_d_pick_facies", None)
        if callable(handler):
            handler()

    def handle_escape(self) -> None:
        """安全退出链（D9 顺序）：退出工具 → 关闭向导 → IDLE。"""
        composite = self._composite
        machine = getattr(composite, "mode_state", None)
        controller = composite.edit_controller
        tool = getattr(controller.tools, "active_tool", None)
        tool_id = str(getattr(tool, "tool_id", "") or "")
        if tool_id and tool_id != "pan":
            controller.activate_tool("pan")
            if machine is not None:
                machine.dispatch(ModeEvent.ESC)
            return
        qc_hub = getattr(composite, "qc_hub", None)
        if qc_hub is not None and qc_hub.isVisible():
            qc_hub.hide()
            if machine is not None:
                machine.dispatch(ModeEvent.ESC)
            return
        if machine is not None:
            machine.dispatch(ModeEvent.ESC)

File: libs/ui_workstation/src/keybinding_manager.cpp
#include "pwb/ui_workstation/keybinding_manager.hpp"

#include <QKeyEvent>

namespace pwb::ui_workstation {

KeybindingHintBar::KeybindingHintBar(QWidget* parent) : QLabel(parent) {
    setObjectName("KeybindingHintBar");
    setAlignment(Qt::AlignmentFlag::AlignCenter);
    setToolTip("编图快捷键随当前模式变化");
}

void KeybindingHintBar::apply_mode(WorkstationMode mode) {
    setText(mode_hint(mode));
}

// ---------------------------------------------------------------------

WorkstationKeyBindingManager::WorkstationKeyBindingManager(
    QObject* parent)
    : QObject(parent) {}

void WorkstationKeyBindingManager::set_hooks(Hooks hooks) {
    hooks_ = std::move(hooks);
}

void WorkstationKeyBindingManager::install_on(QObject* composite,
                                              QObject* canvas,
                                              QObject* qc_hub) {
    if (composite != nullptr) composite->installEventFilter(this);
    if (canvas != nullptr) canvas->installEventFilter(this);
    qc_hub_ = qc_hub;
    if (qc_hub_ != nullptr) qc_hub_->installEventFilter(this);
}

void WorkstationKeyBindingManager::on_tool_state_changed() {
    // 临时平移中的工具切换不是用户模式事件（防模式闪烁）。
    if (pan_restore_tool_.has_value()) return;
    const std::string tool_id =
        hooks_.active_tool_id ? hooks_.active_tool_id() : "";
    if (!tool_id.empty() && tool_id != "pan") {
        if (tool_id != last_tool_id_.value_or("")) {
            last_tool_id_ = tool_id;
            if (hooks_.dispatch_fsm_tool) {
                hooks_.dispatch_fsm_tool(ModeEvent::ToolActivated,
                                         tool_id);
            } else if (hooks_.dispatch_fsm) {
                hooks_.dispatch_fsm(ModeEvent::ToolActivated);
            }
        }
    } else if (last_tool_id_.has_value()) {
        last_tool_id_.reset();
        if (hooks_.dispatch_fsm) {
            hooks_.dispatch_fsm(ModeEvent::ToolDeactivated);
        }
    }
}

KeyEventFacts WorkstationKeyBindingManager::collect_facts(
    KeyCode key, KeyEventType type, bool ctrl, bool shift,
    bool alt) const {
    KeyEventFacts f;
    f.key = key;
    f.type = type;
    f.control = ctrl;
    f.shift = shift;
    f.alt = alt;
    f.text_input_focused =
        hooks_.text_input_focused && hooks_.text_input_focused();
    f.active_tool_id =
        hooks_.active_tool_id ? hooks_.active_tool_id() : "";
    f.active_tool_has_points =
        hooks_.active_tool_has_points && hooks_.active_tool_has_points();
    f.pan_restore_pending = pan_restore_tool_.has_value();
    f.qc_hub_visible =
        hooks_.qc_hub_visible && hooks_.qc_hub_visible();
    f.onion_active = hooks_.onion_active && hooks_.onion_active();
    f.mode_value = hooks_.mode_value ? hooks_.mode_value() : "";
    return f;
}

namespace {

KeyCode key_code_for(int qt_key) {
    switch (qt_key) {
    case Qt::Key_Space:  return KeyCode::Space;
    case Qt::Key_Tab:   return KeyCode::Tab;
    case Qt::Key_Z:     return KeyCode::KeyZ;
    case Qt::Key_X:     return KeyCode::KeyX;
    case Qt::Key_D:     return KeyCode::KeyD;
    case Qt::Key_Escape: return KeyCode::Escape;
    default:            return KeyCode::Other;
    }
}

}  // namespace

bool WorkstationKeyBindingManager::eventFilter(QObject* obj,
                                               QEvent* event) {
    // QC hub Show/Hide → FSM events (never consumed).
    if (obj != nullptr && obj == qc_hub_) {
        if (event->type() == QEvent::Type::Show) {
            if (hooks_.dispatch_fsm) {
                hooks_.dispatch_fsm(ModeEvent::QcHubActivated);
            }
        } else if (event->type() == QEvent::Type::Hide) {
            if (hooks_.dispatch_fsm) {
                hooks_.dispatch_fsm(ModeEvent::QcHubClosed);
            }
        }
        return false;
    }
    if (event->type() == QEvent::Type::KeyPress ||
        event->type() == QEvent::Type::KeyRelease) {
        auto* key_event = static_cast<QKeyEvent*>(event);
        if (key_event->isAutoRepeat() &&
            key_event->key() != Qt::Key_Space) {
            return false;
        }
        // D9：动画平移期间的任何用户按键先打断动画（幂等）。
        if (event->type() == QEvent::Type::KeyPress &&
            hooks_.cancel_smooth_pan) {
            hooks_.cancel_smooth_pan();
        }
        const auto mods = key_event->modifiers();
        const KeyEventFacts facts = collect_facts(
            key_code_for(key_event->key()),
            event->type() == QEvent::Type::KeyPress
                ? KeyEventType::Press
                : KeyEventType::Release,
            mods.testFlag(Qt::KeyboardModifier::ControlModifier),
            mods.testFlag(Qt::KeyboardModifier::ShiftModifier),
            mods.testFlag(Qt::KeyboardModifier::AltModifier));
        const KeyDecision decision = interpret_key(facts);
        if (decision.consumed) {
            run_intent(decision.intent);
        }
        return decision.consumed;
    }
    return false;
}

void WorkstationKeyBindingManager::run_intent(KeyIntent intent) {
    switch (intent) {
    case KeyIntent::BeginTemporaryPan:
        begin_temporary_pan();
        break;
    case KeyIntent::ReleaseTemporaryPan:
        release_temporary_pan();
        break;
    case KeyIntent::CycleSelection:
        if (hooks_.cycle_selection) hooks_.cycle_selection();
        break;
    case KeyIntent::ZoomCenterIn:
        if (hooks_.zoom_by) hooks_.zoom_by(kKeybindingZoomStep);
        break;
    case KeyIntent::ZoomCenterOut:
        if (hooks_.zoom_by) hooks_.zoom_by(1.0 / kKeybindingZoomStep);
        break;
    case KeyIntent::PickFaciesFromSelection:
        if (hooks_.pick_facies) hooks_.pick_facies();
        break;
    case KeyIntent::EscapeCancelGesture:
        if (hooks_.cancel_active_tool) hooks_.cancel_active_tool();
        break;
    case KeyIntent::EscapeEndOnion:
        if (hooks_.end_onion) hooks_.end_onion();
        if (hooks_.dispatch_fsm) hooks_.dispatch_fsm(ModeEvent::Esc);
        break;
    case KeyIntent::EscapeDeactivateTool:
        if (hooks_.activate_tool) hooks_.activate_tool("pan");
        if (hooks_.dispatch_fsm) hooks_.dispatch_fsm(ModeEvent::Esc);
        break;
    case KeyIntent::EscapeCloseQcHub:
        if (hooks_.close_qc_hub) hooks_.close_qc_hub();
        if (hooks_.dispatch_fsm) hooks_.dispatch_fsm(ModeEvent::Esc);
        break;
    case KeyIntent::EscapeDispatchFsm:
        if (hooks_.dispatch_fsm) hooks_.dispatch_fsm(ModeEvent::Esc);
        break;
    default:
        break;
    }
}

void WorkstationKeyBindingManager::begin_temporary_pan() {
    const std::string tool_id =
        hooks_.active_tool_id ? hooks_.active_tool_id() : "";
    if (tool_id.empty() || tool_id == "pan") return;
    pan_restore_tool_ = tool_id;
    if (hooks_.dispatch_fsm) hooks_.dispatch_fsm(ModeEvent::PanHeld);
    if (hooks_.activate_tool) hooks_.activate_tool("pan");
}

void WorkstationKeyBindingManager::release_temporary_pan() {
    const auto restore = pan_restore_tool_;
    pan_restore_tool_.reset();
    if (hooks_.dispatch_fsm) hooks_.dispatch_fsm(ModeEvent::PanReleased);
    if (restore.has_value() && hooks_.activate_tool) {
        hooks_.activate_tool(*restore);
    }
}

void WorkstationKeyBindingManager::handle_escape() {
    const KeyEventFacts facts =
        collect_facts(KeyCode::Escape, KeyEventType::Press,
                      false, false, false);
    run_intent(escape_chain_step(facts));
}

}  // namespace pwb::ui_workstation
